### PR TITLE
[flang][NFC] Added debug output to opt-bufferization pass.

### PR DESCRIPTION
**Pull request by vzakhari**

### CODE REVIEW

**Comment by vzakhari** [comment]
I ran some SPEC and Polyhedron benchmarks on x86, and did not find any noticeable improvements from enabling this option. Of course, the ability to optimize an assignment with allocatable LHS depends on the alias analysis. So we may not see any benefits now, but maybe we will see some later. I would like to keep this option, so that we can quickly run an experiment for an app where we think this might be profitable.

**Review by jeanPerier — APPROVED** [review]
Thanks!

FYI, some compilers actually have user flags to do that since that was the F90 semantics and not everybody was happy with the F2003 semantic change here.

For instance, gfortran has `-fno-realloc-lhs`, ifx has `-assume norealloc_lhs`, and nvfortran has `-Mallocatable=95`.

So we could also make that a  `-fno-realloc-lhs` user option.

File: flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp [hunk 1]
```diff
@@ -462,16 +474,20 @@ ElementalAssignBufferization::findMatch(hlfir::ElementalOp elemental) {
   // the incoming expression
   match.array = match.assign.getLhs();
   mlir::Type arrayType = mlir::dyn_cast<fir::SequenceType>(
-      fir::unwrapPassByRefType(match.array.getType()));
-  if (!arrayType)
+      hlfir::getFortranElementOrSequenceType(match.array.getType()));
+  if (!arrayType) {
```
Comment (tblah) [inline flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp:478]: `getFortranElementOrSequenceType` will return the element type if the passed type is not an array. I think this should be

```suggestion
  if (!mlir::isa<fir::SequenceType>(arrayType)) {
```

**Review by tblah — APPROVED** [review]
LGTM with one nit

**Comment by vzakhari** [comment]
> For instance, gfortran has `-fno-realloc-lhs`, ifx has `-assume norealloc_lhs`, and nvfortran has `-Mallocatable=95`.
> 
> So we could also make that a `-fno-realloc-lhs` user option.

Thank you, Jean! I did not know about this option in other compilers. Then, I guess, the more appropriate way is to not set the `realloc` attribute early during the lowering, when the option is passed.  I will drop the option from this PR, and will only keep the `LLVM_DEBUG` NFC changes (I found them useful when looking at the benchmarks). I will add a new option in a seprate PR.

File: flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp (hunk 1 above)
Comment (vzakhari) [reply flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp:478]: Thanks for the review, Tom! There is a `dyn_cast<fir::SequenceType>` above, which would set `arrayType` to null if it was not an array, so I think it should be okay.  Anyway, I will move this change into another PR.

**Pull request merged**